### PR TITLE
Move macOS Podfile logic into the tool

### LINKED ISSUE #46618: Allow plugins not to support macOS
With the federated model, adding a macos platform to a frontend plugin pubspec.yaml results in a failure. This [line](https://github.com/flutter/flutter/blob/fc05c3738ca64f65484527d7a9a824a0f0bc5706/packages/flutter_tools/templates/cocoapods/Podfile-macos#L38) detects there's a macos plugin, and expects a `macos`. The work around solution is to add a macos folder with a no-op podspec, just as we are doing for [iOS](https://github.com/flutter/flutter/issues/39659) on macos plugins.

### PR DESCRIPTION
**Pull request by jmagman**
## Description

macOS variants of https://github.com/flutter/flutter/pull/59209 and https://github.com/flutter/flutter/pull/59044 and the CocoaPods linking part of https://github.com/flutter/flutter/pull/70224

- Move macOS Podfile logic into the tool so changes can be easily made without the user needing to migrate or regenerate their Podfile.
- Parse `.flutter-plugins-dependencies` instead of the old ` .flutter-plugins` to only embed the `macos`-enabled plugins via CocoaPods.
- Move the Flutter search path and linking behavior out of CocoaPods by generating a "fake" FlutterMacOS.podspec that doesn't point to a real framework (see https://github.com/flutter/flutter/pull/70224 for how the same thing is done for iOS).  This is necessary because plugin podspecs contain `s.dependency 'FlutterMacOS'` and CocoaPods will try to download it from the CoocaPods trunk if the Podfile doesn't `pod 'FlutterMacOS', :path => relative/file`. This has the advantage of making the Podfile.lock stable so `:path: Flutter/ephemeral/.symlinks/flutter/darwin-x64-release` becomes `:path: Flutter/ephemeral/` and doesn't change if you run in different build modes, so can be checked into the user's repo if desired.
- Search path and linking is now handled by adding build settings in the Podfile via the `podhelper.rb`.  Plugins will now link on the frameworks in the artifact cache.  This means the engine artifacts don't need to be symlinked and referred to by CocoaPods, which could have been hairy if any codesigning or (future) ARM thinning actually edited these files.
- Remove now-unused `FLUTTER_FRAMEWORK_DIR` build setting.
- Show warning to tell user to regenerate their Podfile

## Related Issues

Fixes https://github.com/flutter/flutter/issues/46618
https://github.com/flutter/flutter/pull/59044
https://github.com/flutter/flutter/pull/71953
https://github.com/flutter/flutter/pull/70735
One fewer instance of https://github.com/flutter/flutter/issues/48918

## Tests

Updated [plugin_lint_mac](https://github.com/flutter/flutter/pull/71953)

### CODE REVIEW

**Comment by jmagman** [comment]
Rebased on #72372, removed `install! 'cocoapods', :disable_input_output_paths => true` from the podhelper script.

File: dev/benchmarks/macrobenchmarks/macos/Podfile [hunk 1]
```diff
@@ -9,71 +9,32 @@ project 'Runner', {
...
 }
 
-def parse_KV_file(file, separator='=')
-  file_abs_path = File.expand_path(file)
-  if !File.exists? file_abs_path
-    return [];
+def flutter_root
+  generated_xcode_build_settings_path = File.expand_path(File.join('..', 'Flutter', 'ephemeral', 'Flutter-Generated.xcconfig'), __FILE__)
+  unless File.exist?(generated_xcode_build_settings_path)
+    raise "#{generated_xcode_build_settings_path} must exist. If you're running pod install manually, make sure flutter pub get is executed first"
```
Comment (stuartmorgan) [inline dev/benchmarks/macrobenchmarks/macos/Podfile:15]: Nit: "make sure 'flutter pub get' is executed first" (to make parsing out the command part easier)

File: dev/benchmarks/macrobenchmarks/macos/Podfile [hunk 2]
```diff
@@ -9,71 +9,32 @@ project 'Runner', {
...
-    return false;
+  File.foreach(generated_xcode_build_settings_path) do |line|
+    matches = line.match(/FLUTTER_ROOT\=(.*)/)
+    return matches[1].strip if matches
   end
-  File.foreach(file_abs_path) { |line|
-    return true if line =~ /^\s*macos:/
-  }
-  return false
+  raise "FLUTTER_ROOT not found in #{generated_xcode_build_settings_path}. Try deleting Flutter-Generated.xcconfig, then run flutter pub get"
```
Comment (stuartmorgan) [inline dev/benchmarks/macrobenchmarks/macos/Podfile:22]: Same.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM! I love seeing fewer of our internal build details baked into checked-in project files.

File: dev/benchmarks/macrobenchmarks/macos/Podfile (hunk 1 above)
Comment (jmagman) [reply dev/benchmarks/macrobenchmarks/macos/Podfile:15]: Screen Shot 2020-12-17 at 10 34 25 AM

**Pull request merged**